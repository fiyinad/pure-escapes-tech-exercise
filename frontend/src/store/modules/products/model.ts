export interface Product {
    id: number;
    name: string;
    customerPrice: number;
    cost: number;
}

export interface IProductsDomain {
    value: Array<Product>;
    loading: boolean;
    error: boolean;
}

export const initialState: IProductsDomain = {
    value: [],
    loading: false,
    error: false
};
