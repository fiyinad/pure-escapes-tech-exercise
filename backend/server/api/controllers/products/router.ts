import express from 'express';
import controller from './controller';
export default express
  .Router()
  .post('/', controller.create)
  .post('/summary', controller.summary)
  .get('/', controller.all)
  .get('/:id', controller.byId);
