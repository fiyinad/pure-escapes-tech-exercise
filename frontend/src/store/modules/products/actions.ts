import { Product } from "./model";

export const GET_ALL_REQUEST = 'GET_ALL_REQUEST';
export const GET_ALL_SUCCESS = 'GET_ALL_SUCCESS';
export const GET_ALL_FAILURE = 'GET_ALL_FAILURE';
export const GET_ALL = 'GET_ALL';
export const GET_SINGLE = 'GET_SINGLE';

export type GetAllRequestAction = ReturnType<typeof getAllRequestAction>;
export const getAllRequestAction = () => ({
  type: GET_ALL_REQUEST as typeof GET_ALL_REQUEST
});

export type GetAllSuccessAction = ReturnType<typeof getAllSuccessAction>;
export const getAllSuccessAction = () => ({
  type: GET_ALL_SUCCESS as typeof GET_ALL_SUCCESS
});

export type GetAllFailureAction = ReturnType<typeof getAllFailureAction>;
export const getAllFailureAction = () => ({
  type: GET_ALL_FAILURE as typeof GET_ALL_FAILURE
});

export type GetAllAction = ReturnType<typeof getAllAction>;
export const getAllAction = (payload: Array<Product>) => ({
  type: GET_ALL as typeof GET_ALL,
  payload,
});

export type GetSingleAction = ReturnType<typeof getSingleAction>;
export const getSingleAction = (payload: Product) => ({
  type: GET_SINGLE as typeof GET_SINGLE,
  payload 
});

export type ProductsAction =
  | GetAllRequestAction
  | GetAllSuccessAction
  | GetAllFailureAction
  | GetAllAction
  | GetSingleAction;
