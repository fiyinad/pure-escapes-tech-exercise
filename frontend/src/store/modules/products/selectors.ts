import { IProductsDomain } from './model';
import { createSelector } from 'reselect';

export const productsDomain = (state: any): IProductsDomain => state.products;

export const productsSelector = createSelector(
    productsDomain,
    (domain: IProductsDomain): IProductsDomain['value'] => domain.value
);

export const productsLoadingSelector = createSelector(
    productsDomain,
    (domain: IProductsDomain): IProductsDomain['loading'] => domain.loading
);
