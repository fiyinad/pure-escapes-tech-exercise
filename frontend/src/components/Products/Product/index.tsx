import React from 'react';
import { Button, ListItem, ListItemSecondaryAction, ListItemText, Typography } from '@material-ui/core';
import { Product as ProductModel } from '../../../store/modules/products/model';

export interface IProductProps {
    product: ProductModel;
    onAdd: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => void;
    onRemove: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => void;
  }

const Product: React.FC<IProductProps> = props => {
    return (
        <>
            <ListItem>
                <ListItemText>
                    <Typography variant="body1" color='textPrimary'>
                        {props.product?.name}
                    </Typography>
                </ListItemText>
                <ListItemSecondaryAction>
                    <Button 
                        variant="outlined" 
                        aria-label="add" 
                        size="small" 
                        color="primary" 
                        onClick={(e) => props.onAdd(e, props.product)}>
                        Add
                    </Button>
                </ListItemSecondaryAction>
            </ListItem>
        </>
    );
};

export default Product;