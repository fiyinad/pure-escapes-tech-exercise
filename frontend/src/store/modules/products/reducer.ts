import { IProductsDomain, initialState, Product } from './model';
import * as Actions from './actions';

const upsert = (arr: Array<Product>, object: Product) => {
    const i = arr.findIndex(_object => _object.id === object.id);
    if (i > -1) arr[i] = object; 
    else arr.push(object);
    return arr;
}

const productsReducer = (state: IProductsDomain = initialState, action: Actions.ProductsAction) => {
    switch (action.type) {
        case Actions.GET_ALL:
            return { ...state, value: [...action.payload] };

        case Actions.GET_SINGLE:

            return { ...state, value: upsert([...state.value], action.payload) };

        case Actions.GET_ALL_REQUEST:
            return { ...state, loading: true, error: false };

        case Actions.GET_ALL_SUCCESS:
            return { ...state, loading: false, error: false };

        case Actions.GET_ALL_FAILURE:
            return { ...state, loading: false, error: true };

        default:
            return state;
    }
};

export default productsReducer;
