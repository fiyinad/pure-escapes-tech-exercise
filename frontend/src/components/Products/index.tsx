import React, { useState } from 'react';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import axios from 'axios';
import List from '@material-ui/core/List';
import Skeleton from '@material-ui/lab/Skeleton';
import { Button, Divider, ListItem, ListItemSecondaryAction, ListItemText, Modal, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { getAllAction, getAllFailureAction, getAllRequestAction, getAllSuccessAction } from '../../store/modules/products/actions';
import { productsLoadingSelector, productsSelector } from '../../store/modules/products/selectors';
import Product from './Product';
import { Product as ProductModel } from '../../store/modules/products/model';

const useStyles = makeStyles((theme) => ({
    root: {
        width: 360,
        maxWidth: 480,
        backgroundColor: theme.palette.background.paper,
    },
    paper: {
        position: 'absolute',
        width: 400,
        backgroundColor: theme.palette.background.paper,
        border: '2px solid #000',
        boxShadow: theme.shadows[5],
        padding: theme.spacing(2, 4, 3),
    },
    margin: {
        margin: theme.spacing(1),
    },
}));

const Products = () => {
    const classes = useStyles();
    const dispatch = useDispatch();
    const isLoading = useSelector(productsLoadingSelector);
    const [products, setProducts] = useState(useSelector(productsSelector))
    const [openCart, setOpenCart] = useState(false);
    const [cartItems, setCartItems] = useState([] as Array<ProductModel>);
    const modalBody = (
        <div className={classes.paper}>
            <Typography variant="h3" gutterBottom>
                Shopping Cart
            </Typography>
            <Divider />
            <List>
                {cartItems.map((el, idx) => (<ListItem key={idx}>
                    <ListItemText>
                        <Typography variant="body1" color='textPrimary'>
                            {el?.name}
                        </Typography>
                    </ListItemText>
                    <ListItemSecondaryAction>
                        <Button
                            variant="outlined"
                            aria-label="remove"
                            size="small"
                            color="primary"
                            className={classes.margin}
                            onClick={(e) => removeFromCart(e, el)}>
                            Remove
                        </Button>
                    </ListItemSecondaryAction>
                </ListItem>))}
            </List>
        </div>
    );

    useEffect(() => {
        const fetchData = async () => {
            return await axios.get('http://localhost:3000/api/v1/products');
        };
        dispatch(getAllRequestAction());
        fetchData().then(response => {
            dispatch(getAllAction(response?.data));
            dispatch(getAllSuccessAction());
            setProducts(response?.data as Array<ProductModel>);
        }).catch(() => dispatch(getAllFailureAction()));
    }, [dispatch]);

    const addToCart = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => {
        e.stopPropagation();
        cartItems.push(item);
    };

    const removeFromCart = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => {
        e.stopPropagation();
        const tmpCartItems = [...cartItems];
        const index = tmpCartItems.indexOf(item);
        if (index > -1) {
            tmpCartItems.splice(index, 1);
        }
        setCartItems(tmpCartItems);
    }

    const toggleCart = () => {
        setOpenCart(prevProp => !prevProp);
    }

    const handleClose = () => {
        setOpenCart(false);
    };

    return (
        <div className={classes.root}>
            {isLoading && <Skeleton> <List /></Skeleton>}
            {
                isLoading === false &&
                products?.length > 0 && (
                    <>
                        <List>
                            {products.map((el, idx) => <Product key={idx} onAdd={addToCart} onRemove={removeFromCart} product={el} />)}
                        </List>
                        <Button onClick={toggleCart}>{openCart ? 'Hide' : 'Show'} Cart</Button>
                    </>
                )

            }
            {<Modal
                open={openCart}
                onClose={handleClose}
                aria-labelledby="simple-modal-title"
                aria-describedby="simple-modal-description"
                style={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                {modalBody}
            </Modal>}
        </div>
    );
};

export default Products;