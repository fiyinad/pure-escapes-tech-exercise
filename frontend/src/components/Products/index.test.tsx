import React from 'react';
import * as ReactRedux from 'react-redux'
import axios from 'axios';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Products from './index';
import { initialState } from '../../store/modules/products/model';

describe('Products', () => {
    const mockXXXFn = jest.fn();
    const mockUseDispatch =
        jest.spyOn(ReactRedux, 'useDispatch').mockReturnValue(mockXXXFn);
    const mockUseSelector =
        jest.spyOn(ReactRedux, 'useSelector').mockReturnValue(initialState);
    const mockAxios = jest.spyOn(axios, 'get');

    afterEach(() => {
        mockUseDispatch.mockClear();
        mockUseSelector.mockClear();
        mockAxios.mockClear();
    })

    it('Renders without errors in default state', () => {
        // arrange, act
        const component = shallow(<Products />);

        // assert
        expect(toJson(component)).toMatchSnapshot();
    });

    it('Renders without errors in loading state', () => {
        // arrange
        const mockedState = {
            ...initialState,
            loading: true
        }
        mockUseSelector.mockReturnValueOnce(mockedState.loading); // mock productsLoadingSelector

        // act
        const component = shallow(<Products />);

        // assert
        expect(toJson(component)).toMatchSnapshot();
    });

    it('Renders without errors with pre-loaded state', () => {
        // arrange
        const payload = [
            { "id": 1, "name": "Soup", "customerPrice": 199, "cost": 186 },
            { "id": 2, "name": "Bread", "customerPrice": 87, "cost": 21 },
            { "id": 3, "name": "Cheese", "customerPrice": 275, "cost": 234 },
            { "id": 4, "name": "Milk", "customerPrice": 67, "cost": 61 }];
        const mockedState = {
            ...initialState,
            value: payload
        };
        mockAxios.mockImplementationOnce(() => Promise.resolve(payload));
        mockUseSelector.mockReturnValueOnce(mockedState.loading); // mock productsLoadingSelector
        mockUseSelector.mockReturnValueOnce(mockedState.value); // mock productsSelector

        // act
        const component = shallow(<Products />);
        component.update();

        // assert
        expect(toJson(component)).toMatchSnapshot();
    });
});