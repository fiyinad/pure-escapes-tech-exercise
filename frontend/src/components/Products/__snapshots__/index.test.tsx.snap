// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Renders without errors in default state 1`] = `
<div
  className="makeStyles-root-1"
>
  <ForwardRef(Modal)
    aria-describedby="simple-modal-description"
    aria-labelledby="simple-modal-title"
    onClose={[Function]}
    open={false}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="makeStyles-paper-2"
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h3"
      >
        Shopping Cart
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(List)) />
    </div>
  </ForwardRef(Modal)>
</div>
`;

exports[`Products Renders without errors in loading state 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Skeleton))>
     
    <WithStyles(ForwardRef(List)) />
  </WithStyles(ForwardRef(Skeleton))>
  <ForwardRef(Modal)
    aria-describedby="simple-modal-description"
    aria-labelledby="simple-modal-title"
    onClose={[Function]}
    open={false}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="makeStyles-paper-2"
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h3"
      >
        Shopping Cart
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(List)) />
    </div>
  </ForwardRef(Modal)>
</div>
`;

exports[`Products Renders without errors with pre-loaded state 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(List))>
    <Product
      key="0"
      onAdd={[Function]}
      onRemove={[Function]}
      product={
        Object {
          "cost": 186,
          "customerPrice": 199,
          "id": 1,
          "name": "Soup",
        }
      }
    />
    <Product
      key="1"
      onAdd={[Function]}
      onRemove={[Function]}
      product={
        Object {
          "cost": 21,
          "customerPrice": 87,
          "id": 2,
          "name": "Bread",
        }
      }
    />
    <Product
      key="2"
      onAdd={[Function]}
      onRemove={[Function]}
      product={
        Object {
          "cost": 234,
          "customerPrice": 275,
          "id": 3,
          "name": "Cheese",
        }
      }
    />
    <Product
      key="3"
      onAdd={[Function]}
      onRemove={[Function]}
      product={
        Object {
          "cost": 61,
          "customerPrice": 67,
          "id": 4,
          "name": "Milk",
        }
      }
    />
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Button))
    onClick={[Function]}
  >
    Show
     Cart
  </WithStyles(ForwardRef(Button))>
  <ForwardRef(Modal)
    aria-describedby="simple-modal-description"
    aria-labelledby="simple-modal-title"
    onClose={[Function]}
    open={false}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="makeStyles-paper-2"
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h3"
      >
        Shopping Cart
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(List)) />
    </div>
  </ForwardRef(Modal)>
</div>
`;
